CONTRACTS: store contracts in dedicated symbols

tautschnig opened the pull request:
Contracts can now be placed on declarations or definitions, and result
in a fresh symbol being generated. That new symbol has its name prefixed
with "contracts::" to place them in a pseudo-namespace, and has
"is_property" set (which was an otherwise unused symbolt flag).

Co-authored-by: Remi Delmas <delmasrd@amazon.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
CONTRACTS: store contracts in dedicated symbols

Contracts can now be placed on declarations or definitions, and result
in a fresh symbol being generated. That new symbol has its name prefixed
with "contracts::" to place them in a pseudo-namespace, and has
"is_property" set (which was an otherwise unused symbolt flag).

Co-authored-by: Remi Delmas <delmasrd@amazon.com>

remi-delmas-3000 commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -744,6 +760,48 @@ void c_typecheck_baset::typecheck_declaration(
+        contract.base_name = new_symbol.name;
+        contract.pretty_name = new_symbol.name;
+        contract.is_property = true;
+        contract.type = code_type;
```
Will the contract symbol also contain the function body if one was given by the user ? 

If possible I would like the contract symbol to have an empty body but to keep the contract parameters in the symbol table, because in a second step (goto-convert maybe ?) I would like to generate an executable body from the contract clauses.

tautschnig replied:
No, the contract symbol _at present_ really just has the type information including the contract annotation. We could obviously evolve that.

I'm not sure whether symbols for the contract parameters are required to remain in the symbol table: even the type signature already has the parameters with their names and their types, so we can easily create symbols from them if need be.

kroening replied:
Use a binding instead!

tautschnig replied:
> Use a binding instead!

For all that I can tell this requires some preparing work. https://github.com/diffblue/cbmc/pull/6811 will be a starting point.

tautschnig replied:
I have now added a commit that wraps contracts in a `lambda_exprt` and uses `.application()` instead of resorting to `replace_symbolt`.

remi-delmas-3000 commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -744,6 +760,48 @@ void c_typecheck_baset::typecheck_declaration(
+        // checking the contract
+        if(new_symbol.value.is_nil())
+        {
+          for(const auto &parameter : code_type.parameters())
+          {
+            const irep_idt &identifier = parameter.get_identifier();
+
+            symbol_tablet::symbolst::const_iterator p_s_it =
+              symbol_table.symbols.find(identifier);
+            if(p_s_it != symbol_table.symbols.end())
+              symbol_table.erase(p_s_it);
```
Here I would rather replace the contract parameters with fresh symbols (maybe just by prefixing the function parameters with a `contract::`  prefix, and also rewrite the contract clauses with these new parameters using a `replacet`. 

Later we would be able to generate a body for the contract without symbol clashes between function and contract symbols and make the contract fully self-contained.

tautschnig replied:
I believe we could still choose to create those fresh symbols if and when needed: all the required information still is in the contract symbol.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -727,6 +727,29 @@ code_contractst::create_ensures_instruction(
+  }
+
+  const auto &type = to_code_with_contract_type(contract_sym->type);
+  if(type != function_symbol.type)
```
How can this equality work if the function symbol had its contract clauses removed but the contract symbol still has them ? 

Would the check still work if in addition we removed the function body from the contract and renamed contract parameters ? 

Maybe we could just check that the number of arguments and their types are equal.

tautschnig replied:
This perhaps warrants an extra comment: the comparison works, because all the extra information is stored in comment-elements of the ireps: both the parameter identifiers as well as all the contract clauses have `ID_C_...` keys in the `irept`. Therefore, equality (but not full equality!) works out. While the body isn't there as indicated in the earlier comment, it wouldn't affect this comparison either way for it's just a comparison of the types, not the symbols' values.

remi-delmas-3000 commented on src/linking/remove_internal_symbols.cpp:
```diff
@@ -182,6 +183,8 @@ void remove_internal_symbols(
         get_symbols(ns, symbol, exported);
       }
+      else if(is_contract)
+        get_symbols(ns, symbol, exported);
```
If a function `foo` is file_local, is also `contract::foo` file_local ? 

What if some of the contract's clauses use file_local symbols in preconditions or post conditions or assigns clause ?

tautschnig replied:
No, the contract does not currently have any flags (other than `is_property` to mark it as a contract) set, and will therefore not be considered "file_local."

Your call-out about file_local symbols is a good one as the linker could choose to rename them in case of a conflict, at which point the contract would no longer refer to the correct symbol. I'll try to come up with some code that tracks this renaming and adjusts the contract accordingly.

remi-delmas-3000 replied:
Hi again,
I experimented with the following:


```c
// foo.h
#include <stdlib.h>
void foo(char *arr, size_t size);
```

```c
// foo.c
#include "foo.h"
void foo(char *arr, size_t size) {
    if(arr && size > 0)
        arr[0] = 0;
}
```

```c
// foo_contract.c
#include <stdlib.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```

```c
// main.c
#include "foo.h"
void main() {
    size_t size = 10;
    char *arr = malloc(size);
    foo(arr, size);
}
```

and then 

```bash
▶goto-cc foo_contract.c foo.c main.c  -o project.o && goto-instrument project.o project-mod.o --enforce-contract foo && cbmc project-mod.o

** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.2] line 4 Check that arr[(signed long int)0] is assignable: SUCCESS
[foo.assigns.1] file foo_contract.c line 4 Check that arr[(signed long int)0] is valid when arr != ((char *)NULL) && size > (unsigned long int)0: SUCCESS

** 0 of 4 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

But if I include `"foo.h"` in `"foo_contract.h"` like so

```c
// foo_contract.c
#include <stdlib.h>
#include <foo.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```


Then it seems like the contract clauses are lost.

```bash
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.1] line 4 Check that arr[(signed long int)0] is assignable: FAILURE

** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```

tautschnig replied:
Good catch, we need to maintain the contract in `typecheck_redefinition_non_type` even when the new symbol does not provide the function definition. I'll add this case and will try to include a test.

tautschnig replied:
Thanks @remi-delmas-3000, I believe I have now fixed this problem (I have also extended the test suite to capture this scenario).

remi-delmas-3000 reviewed (commented):
It would be nice to make the contract completely self-contained by introducing fresh/prefixed names for the parameters, rewriting the contract clauses with the fresh names, and removing the body if any.

tautschnig committed:
Code contracts: remove useless skip statement

All that was used was its source location, which we can obtain in a
different way.

tautschnig committed:
Wrap contract expressions in a lambda

This makes sure contracts are self-contained in that any parameter names
that may have only existed in the scope of a function declaration are
now bound variables.

tautschnig commented:
@kroening @remi-delmas-3000 I have put the additional changes to make use of `lambda_exprt` in a commit of its own, but it should now be working.

feliperodri commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -370,7 +373,15 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
     if(old_ct.has_ellipsis() && !new_ct.has_ellipsis())
       old_ct=new_ct;
     else if(!old_ct.has_ellipsis() && new_ct.has_ellipsis())
+    {
+      if(to_code_with_contract_type(new_ct).has_contract())
+      {
+        error().source_location = new_symbol.location;
+        error() << "code contract on incomplete function re-declaration" << eom;
+        throw 0;
+      }
```
nitpick: can we cover this error with a regression test?

tautschnig replied:
Done!

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -734,6 +733,29 @@ code_contractst::create_ensures_instruction(
+  const auto &type = to_code_with_contract_type(contract_sym->type);
+  if(type != function_symbol.type)
+  {
+    throw invalid_input_exceptiont(
+      "Contract of '" + function_str + "' has different signature.");
+  }
```
nitpick: could we cover this with a regression test?

tautschnig replied:
Done!

feliperodri reviewed (approved):
LGTM, thanks @tautschnig!

SaswatPadhi commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -734,6 +733,29 @@ code_contractst::create_ensures_instruction(
+get_contract(const irep_idt &function, const namespacet &ns)
+{
+  const std::string function_str = id2string(function);
+  const symbolt &function_symbol = ns.lookup(function);
```
[nit] could use `auto` here
```suggestion
  const auto function_str = id2string(function);
  const auto &function_symbol = ns.lookup(function);
```

SaswatPadhi commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1445,38 +1475,34 @@ void code_contractst::add_contract_check(
-  goto_programt tmp_skip;
-  goto_programt::targett skip =
-    tmp_skip.add(goto_programt::make_skip(ensures.source_location()));
-
```
Thanks for cleaning this up. We don't seem to be using this `skip` target any more.
While you're at it, could you please also update the pseudocode in the comments above to reflect the current instrumentation?

SaswatPadhi reviewed (approved):
LGTM overall; couple of minor comments below.

feliperodri commented:
@chris-ryder, @kroening, @martin-cs, and/or @peterschrammel, could we get one more review on this PR? It's missing code owner approval.
